File: csharp/src/ConsoleForum.App/Core/ConsoleForumCommandManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Reflection.Metadata.Ecma335;
using System.Text;
using System.Threading;
using System.Windows.Input;
using ConsoleForum.App.Commands.Attributes;
using ConsoleForum.App.Common;
using ConsoleForum.App.IO;
using ConsoleForum.App.Models;
using ConsoleForum.App.Models.Enums;
using ICommand = ConsoleForum.App.Commands.ICommand;

namespace ConsoleForum.App.Core
{
    public class ConsoleForumCommandManager : ICommandManager
    {
        private readonly IInputReader consoleReader;

        private readonly IOutputWriter consoleWriter;

        private IPrincipal currentlyLoggedInUser;

        public ConsoleForumCommandManager(IInputReader consoleReader, IOutputWriter consoleWriter)
        {
            this.consoleReader = consoleReader;
            this.consoleWriter = consoleWriter;
            this.currentlyLoggedInUser = new Principal();
        }

        private bool IsLoggedIn() => currentlyLoggedInUser.User != null;

        private bool IsAdmin() => this.IsLoggedIn() && this.currentlyLoggedInUser.User.Role.Equals(UserRole.Admin);

        private bool IsModerator() => this.IsLoggedIn() && this.currentlyLoggedInUser.User.Role.Equals(UserRole.Moderator);

        private string GetAppropriateHeader()
        {
            if (this.IsAdmin()) return string.Format(Constants.AdminIntroductionMessage, this.currentlyLoggedInUser.User.Username, new string(' ', 37 - this.currentlyLoggedInUser.User.Username.Length));
            if (this.IsModerator()) return string.Format(Constants.ModeratorIntroductionMessage, this.currentlyLoggedInUser.User.Username, new string(' ', 43 - this.currentlyLoggedInUser.User.Username.Length));
            if (this.IsLoggedIn()) return string.Format(Constants.UserIntroductionMessage, this.currentlyLoggedInUser.User.Username, new string(' ', 53 - this.currentlyLoggedInUser.User.Username.Length));
            return Constants.GuestIntroductionMessage;
        }

        private void ClearConsole() => Console.Clear();

        private void PrintIntro()
        {
            this.consoleWriter.WriteLine(Constants.ConsoleForumIntroBanner);
            this.consoleWriter.WriteLine(Constants.ConsoleForumOutputEmptyLine);
            this.consoleWriter.WriteLine(this.GetAppropriateHeader());
            this.consoleWriter.WriteLine(Constants.ConsoleForumOutputPrefix);
        }

        private ICommand InstanceCommand(string input)
        {
            List<string> inputParams = input.Split("|").ToList();
            string commandName = inputParams[0].ToLower();
            inputParams.RemoveAt(0);

            var commandType = Assembly.GetEntryAssembly()
                ?.GetTypes()
                .Where(type => typeof(ICommand).IsAssignableFrom(type))
                .FirstOrDefault(type => type.Name.Replace("Command", string.Empty).ToLower() == commandName);

            if (commandType != null)
            {
                ICommand command = (ICommand)Activator.CreateInstance(commandType);

                command.ConsoleReader = this.consoleReader;
                command.ConsoleWriter = this.consoleWriter;
                command.Parameters = inputParams;
                command.Principal = this.currentlyLoggedInUser;

                return command;
            }

            return null;
        }

        private bool IsAuthorized(ICommand command)
        {
            var authAttributes = command
                .GetType()
                .GetCustomAttributes()
                .Where(attribute => attribute.GetType().Name.Contains("CommandAttribute"))
                .ToList();
                
            if (!authAttributes.Any())
            {
                return true;
            }
            else
            {
                return 
                    (this.IsLoggedIn() && authAttributes
                    .Any(attribute => attribute.GetType().Name.Contains(this.currentlyLoggedInUser.User.Role.ToString())))
                    || (!this.IsLoggedIn() && authAttributes
                            .Any(attribute => attribute.GetType() == typeof(GuestCommandAttribute)));
            }
        }

        public void Reset()
        {
            this.ClearConsole();
            this.PrintIntro();
        }

        public void HandleInput()
        {
            this.consoleWriter.Write(Constants.ConsoleForumOutputPrefix);
            string input = this.consoleReader.ReadLine();

            ICommand command = this.InstanceCommand(input);

            if (command == null)
            {
                this.consoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                this.consoleWriter.ErrorLine("Unsupported command...");
                Thread.Sleep(Constants.NotificationDelay);
                this.Reset();
            }
            else
            {
                if (!this.IsAuthorized(command))
                {
                    this.consoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                    this.consoleWriter.ErrorLine("You have no access to this command.");
                    Thread.Sleep(Constants.NotificationDelay);
                    this.Reset();
                }
                else
                {
                    command.Execute();
                    if (command.ResetsConsole) this.Reset();
                }
            }
        }
    }
}

File: java/src/org/sanity/consoleForum/core/ConsoleForumCommandManager.java
package org.sanity.consoleForum.core;

import org.sanity.consoleForum.commands.Attributes.AdminCommand;
import org.sanity.consoleForum.commands.Attributes.GuestCommand;
import org.sanity.consoleForum.commands.Attributes.ModeratorCommand;
import org.sanity.consoleForum.commands.Attributes.UserCommand;
import org.sanity.consoleForum.commands.Command;
import org.sanity.consoleForum.common.Constants;
import org.sanity.consoleForum.database.EntityManager;
import org.sanity.consoleForum.io.InputReader;
import org.sanity.consoleForum.io.OutputWriter;
import org.sanity.consoleForum.models.enums.UserRole;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class ConsoleForumCommandManager implements CommandManager {
    private final InputReader consoleReader;

    private final OutputWriter consoleWriter;

    private final EntityManager entityManager;

    private Principal currentlyLoggedInUser;

    public ConsoleForumCommandManager(InputReader consoleReader, OutputWriter consoleWriter, EntityManager entityManager) {
        this.consoleReader = consoleReader;
        this.consoleWriter = consoleWriter;
        this.entityManager = entityManager;
        this.currentlyLoggedInUser = new PrincipalImpl();
    }

    // TODO: METHODS TO CHECK IF THE CURRENT USER IS LOGGED IN AND IS ADMIN OR MODERATOR
    private boolean isLoggedIn() {
        return false; // TODO: IMPLEMENT ME ...
    }

    private boolean isAdmin() {
        return false; // TODO: IMPLEMENT ME ...
    }

    private boolean isModerator() {
        return false; // TODO: IMPLEMENT ME ...
    }

    private String getAppropriateHeader() {
        if (this.isAdmin()) return String.format(Constants.ADMIN_INTRODUCTION_MESSAGE, this.currentlyLoggedInUser.getUser().getUsername(), " ".repeat(37 - this.currentlyLoggedInUser.getUser().getUsername().length()));
        if (this.isModerator()) return String.format(Constants.MODERATOR_INTRODUCTION_MESSAGE, this.currentlyLoggedInUser.getUser().getUsername(), " ".repeat(43 - this.currentlyLoggedInUser.getUser().getUsername().length()));
        if (this.isLoggedIn()) return String.format(Constants.USER_INTRODUCTION_MESSAGE, this.currentlyLoggedInUser.getUser().getUsername(), " ".repeat(53 - this.currentlyLoggedInUser.getUser().getUsername().length()));
        return Constants.GUEST_INTRODUCTION_MESSAGE;
    }

    private void clearConsole() {
        try {
            if (System.getProperty("os.name").contains("Windows"))
                new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
            else
                Runtime.getRuntime().exec("clear");
        } catch (IOException | InterruptedException ignored) { }
    }

    private void printIntro() {
        try {
            this.consoleWriter.writeLine(Constants.CONSOLE_FORUM_INTRO_BANNER);
            this.consoleWriter.writeLine(Constants.CONSOLE_FORUM_OUTPUT_EMPTY_LINE);
            this.consoleWriter.writeLine(this.getAppropriateHeader());
            this.consoleWriter.writeLine(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // THIS CREATES A COMMAND OBJECT FROM A CLASS BASED ON INPUT
    private Command instanceCommand(String input) throws ClassNotFoundException, NoClassDefFoundError, NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
        List<String> inputParams = Arrays.stream(input.split("\\|")).collect(Collectors.toList());
        String commandName = inputParams.get(0);
        inputParams.remove(0);

        Class commandType = null;

        commandType = Class.forName("org.sanity.consoleForum.commands." + commandName + "Command");

        if (commandType != null) {
            Command command = null;

            command = (Command) commandType.getDeclaredConstructor().newInstance();
            command.setConsoleReader(this.consoleReader);
            command.setConsoleWriter(this.consoleWriter);
            command.setParameters(inputParams);
            command.setPrincipal(this.currentlyLoggedInUser);
            command.setEntityManager(this.entityManager);

            return command;
        }

        return null;
    }

    private boolean isAuthorized(Command command) {
        Class<?> commandClass = command.getClass();

        if (
                (this.currentlyLoggedInUser.getUser() == null && !commandClass.isAnnotationPresent(GuestCommand.class)) ||
                        (this.currentlyLoggedInUser.getUser() != null && this.currentlyLoggedInUser.getUser().getRole() == UserRole.USER
                                && !commandClass.isAnnotationPresent(UserCommand.class)) ||
                        (this.currentlyLoggedInUser.getUser() != null && this.currentlyLoggedInUser.getUser().getRole() == UserRole.MODERATOR
                                && !commandClass.isAnnotationPresent(ModeratorCommand.class)) ||
                        (this.currentlyLoggedInUser.getUser() != null && this.currentlyLoggedInUser.getUser().getRole() == UserRole.ADMIN
                                && !commandClass.isAnnotationPresent(AdminCommand.class))) {
            return false;
        }

        return true;
    }

    public void reset() {
        this.clearConsole();
        this.printIntro();
    }

    public void handleInput() throws IOException, NoClassDefFoundError, InterruptedException, ClassNotFoundException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
        String input = this.consoleReader.readLine();

        Command command = this.instanceCommand(input);

        if (command == null) {
            this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
            this.consoleWriter.errorLine("Unsupported command...", null);
            Thread.sleep(Constants.NOTIFICATION_DELAY);
            this.reset();
        } else {
            if (!this.isAuthorized(command)) {
                this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
                this.consoleWriter.errorLine("You have no access to this command.", null);
                Thread.sleep(Constants.NOTIFICATION_DELAY);
                this.reset();
            } else {
                command.execute();
                if (command.getResetsConsole()) this.reset();
            }
        }
    }
}
